### PR TITLE
feat(2767): Get stage workflowGraph [2]

### PR DESCRIPTION
**Pull request by tkyi**
## Context

Stages might want to always run initialization or cleanups needed before jobs in a stage can be executed.

## Objective

This PR get workflowGraph for stages.

## References

Related to https://github.com/screwdriver-cd/screwdriver/issues/2767

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: lib/getWorkflow.js [hunk 1]
```diff
@@ -158,23 +250,51 @@ const calculateEdges = async (jobs, triggerFactory, externalDownstreamOrs, exter
...
-                    edges.push({ src: filterNodeName(src), dest: jobName });
+                    const [source] = src.split(':');
+
+                    if (stageObj) {
+                        let stageEdge = { src: filterNodeName(src), dest: jobName };
+
+                        // Remove stage prefix when src is not stage setup
+                        if (!stageObj.startFrom) {
+                            stageEdge = { src: filterNodeName(source), dest: jobName };
+                        }
```
Comment (sagar1312) [inline lib/getWorkflow.js:274]: This block may not be needed. What is the use case which needs this?

I believe when the condition satisfies, `src` and `source` would be identical.
```suggestion
                        // Remove stage prefix when src is not stage setup
                        if (!stageObj.startFrom) {
                            stageEdge = { src: filterNodeName(source), dest: jobName };
                        }
```

**Review by sagar1312 — CHANGES_REQUESTED** [review]

File: lib/getWorkflow.js [hunk 2]
```diff
@@ -158,23 +250,51 @@ const calculateEdges = async (jobs, triggerFactory, externalDownstreamOrs, exter
...
                     const obj = { src, dest: jobName };
 
                     if (isJoin) {
                         obj.join = true;
                     }
-                    edges.push(obj);
+
+                    // Skip adding edge when src is stage setup
+                    if (stageObj) {
+                        if (!stageObj.startFrom) {
```
[the remark this answers is not in the document]
Comment (tkyi) [reply lib/getWorkflow.js:292]: yes

File: lib/getWorkflow.js (hunk 1 above)
Comment (tkyi) [reply lib/getWorkflow.js:274]: I think there's redundant edges if we don't have this

File: lib/getFlattenedWorkflow.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * Flatten workflowGraphs (event and stage)
+ * @method flattenWorkflow
+ * @param  {Object}  workflowGraph  Graph representation of workflow
+ * @return {Boolean}                True if a cycle exists anywhere in the workflow
+ */
+const getFlattenedWorkflow = (workflowGraph, stageWorkflows) => {
+    const flattenedWorkflow = { edges: [], nodes: [] };
+
+    console.log('----workflowGraph: ', workflowGraph);
+    console.log('----stageWorkflows: ', stageWorkflows);
```
[the remark this answers is not in the document]
Comment (tkyi) [reply lib/getFlattenedWorkflow.js:14]: yes

File: lib/getWorkflow.js [hunk 4]
```diff
@@ -30,6 +33,66 @@ const buildExternalNodes = async (children, nodes, triggerFactory) => {
...
+ */
+const handleDisplayName = (jobs, jobName, node) => {
+    // Add display name
+    const displayName = hoek.reach(jobs[jobName], 'annotations>screwdriver.cd/displayName', {
+        separator: '>',
+        default: undefined
+    });
+
+    if (displayName !== undefined) {
+        node.displayName = displayName;
```
[the remark this answers is not in the document]
Comment (tkyi) [reply lib/getWorkflow.js:50]: this was just a refactor of previous code: https://github.com/screwdriver-cd/workflow-parser/pull/43/files#diff-fb9c28c92e1664b0dadebca039a215ab6f0b756fb9362b39efec7e1baf67e16cL87
I can look into this after we get the stage functionality working

File: lib/getWorkflow.js [hunk 5]
```diff
@@ -54,14 +117,36 @@ const calculateNodes = async (jobs, triggerFactory, externalDownstreamOrs, exter
...
+            } else {
+                nodes.add(jobName);
+            }
 
             // upstream nodes
             if (Array.isArray(jobs[jobName].requires)) {
-                jobs[jobName].requires.forEach(n => nodes.add(filterNodeName(n)));
+                jobs[jobName].requires.forEach(n => {
+                    // Stage node
+                    if (/^(?:~)?(stage@)/.test(n)) {
```
[the remark this answers is not in the document]
Comment (tkyi) [reply lib/getWorkflow.js:142]: will add a check to prevent duplicates

File: lib/getWorkflow.js [hunk 6]
```diff
@@ -76,19 +161,25 @@ const calculateNodes = async (jobs, triggerFactory, externalDownstreamOrs, exter
...
-    return [...nodes].map(name => {
+    // Create stage nodes in main workflowGraph
+    Object.keys(stagesDict).forEach(stageName => {
+        nodes.add(`~${STAGE_PREFIX}${stageName}`);
+
+        // Create stage setup/teardown node if they do not exist
+        stagesDict = addSetupTeardownNode(stagesDict, stageName, 'setup');
+        stagesDict = addSetupTeardownNode(stagesDict, stageName, 'teardown');
+    });
+
```
[the remark this answers is not in the document]
Comment (tkyi) [reply lib/getWorkflow.js:172]: If someone uses workflow-parser without using config-parser, setup/teardown might not be there. But I can remove.

We still need line 166 to add the stage node(ie, `~stage@deploy`) to the pipelineWorkflow nodes

File: lib/getWorkflow.js (hunk 1 above)
Comment (tkyi) [reply lib/getWorkflow.js:270]: removed

File: lib/getWorkflow.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tkyi) [reply lib/getWorkflow.js]: fixed

File: test/lib/getWorkflow.test.js [hunk 7]
```diff
@@ -21,72 +21,179 @@ describe('getWorkflow', () => {
...
+                { name: '~stage@test' }
+            ],
+            edges: [
+                { src: '~commit', dest: 'foo' },
+                { src: 'foo', dest: 'A' },
+                { src: 'foo', dest: 'B', join: true },
+                { src: 'stage@other', dest: 'B', join: true },
+                { src: 'A', dest: 'stage@deploy' },
+                { src: 'foo', dest: 'stage@test' }
+            ]
```
Comment (sagar1312) [inline test/lib/getWorkflow.test.js:150]: Actual result (below) is not matching the expectation
```
[
  {
    "src": "~commit",
    "dest": "foo"
  },
  {
    "src": "foo",
    "dest": "A"
  },
  {
    "src": "foo",
    "dest": "B",
    "join": true
  },
  {
    "src": "stage@other",
    "dest": "B",
    "join": true
  }
]
```

File: test/lib/getWorkflow.test.js [hunk 8]
```diff
@@ -21,72 +21,179 @@ describe('getWorkflow', () => {
...
+            other: {
+                nodes: [
+                    { name: 'stage@other:teardown' },
+                    { name: 'main' },
+                    { name: 'stage@other:setup' },
+                    { name: 'publish' }
+                ],
+                edges: [
+                    { src: 'stage@other:setup', dest: 'main' },
+                    { src: 'main', dest: 'publish' }
```
Comment (sagar1312) [inline test/lib/getWorkflow.test.js:162]: Shouldn't there be an edge from leaf nodes ([`publish`]) to `teardown`?
```suggestion
                    { src: 'stage@other:setup', dest: 'main' },
                    { src: 'main', dest: 'publish' },
                     { src: 'publish', dest: 'stage@other:teardown' }
```

Comment (sagar1312) [reply test/lib/getWorkflow.test.js:162]: Also the actual result (below) is not matching the expectation
```
[
  {
    "src": "main",
    "dest": "publish"
  }
]
```

File: test/lib/getWorkflow.test.js [hunk 9]
```diff
@@ -21,72 +21,179 @@ describe('getWorkflow', () => {
...
+            deploy: {
+                nodes: [
+                    { name: 'C' },
+                    { name: 'stage@deploy:setup' },
+                    { name: 'D' },
+                    { name: 'stage@deploy:teardown' }
+                ],
+                edges: [
+                    { src: 'stage@deploy:setup', dest: 'C' },
+                    { src: 'C', dest: 'D' }
```
[the remark this answers is not in the document]
Comment (sagar1312) [reply test/lib/getWorkflow.test.js:174]: Also the actual result (below) is not matching the expectation
```
[
  {
    "src": "C",
    "dest": "D"
  }
]
```

File: test/lib/getWorkflow.test.js [hunk 10]
```diff
@@ -21,72 +21,179 @@ describe('getWorkflow', () => {
...
+                ]
+            },
+            test: {
+                nodes: [{ name: 'E' }, { name: 'stage@test:setup' }, { name: 'stage@test:teardown' }],
+                edges: [{ src: 'stage@test:setup', dest: 'E' }]
```
[the remark this answers is not in the document]
Comment (sagar1312) [reply test/lib/getWorkflow.test.js:179]: Also the actual result (below) is not matching the expectation
```
[
  {
    "src": "foo",
    "dest": "stage@test:setup"
  }
]
```

**Review by sagar1312 — CHANGES_REQUESTED** [review]

File: lib/getWorkflow.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (sagar1312) [reply lib/getWorkflow.js]: >We still need line 166 to add the stage node(ie, ~stage@deploy) to the pipelineWorkflow nodes

👍 

>If someone uses workflow-parser without using config-parser, setup/teardown might not be there. But I can remove.

Yes, we should remove.
Whoever is calling `worflow-parser` should make sure they pass valid input. `worflow-parser`'s responsibility is to build workflow graph based on the input. It should not be creating `nodes` for jobs (`setup`/`teardown`) which do not exist.

File: test/lib/getWorkflow.test.js [hunk 11]
```diff
@@ -21,72 +21,179 @@ describe('getWorkflow', () => {
...
 
-    it('should handle logical OR requires', async () => {
-        const result = await getWorkflow(
-            {
-                jobs: {
-                    foo: { requires: ['~commit'] },
-                    A: { requires: ['foo'] },
-                    B: { requires: ['foo'] },
-                    C: { requires: ['~A', '~B', '~sd@1234:foo'] }
+    it('should handle stage requires', async () => {
```
Comment (sagar1312) [inline test/lib/getWorkflow.test.js:91]: This test is focused on `AND` use-cases.

Per https://github.com/screwdriver-cd/data-schema/pull/510#discussion_r1260398901, for initial release, we will be supporting only `OR`. Can we have similar test to cover `OR` use-cases.

File: lib/getWorkflow.js [hunk 12]
```diff
@@ -166,15 +275,36 @@ const calculateEdges = async (jobs, triggerFactory, externalDownstreamOrs, exter
...
                     const obj = { src, dest: jobName };
 
                     if (isJoin) {
                         obj.join = true;
                     }
-                    edges.push(obj);
+
+                    // Skip adding edge when src is stage setup
+                    if (stageObj) {
+                        if (!src.endsWith(STAGE_START)) {
```
Comment (sagar1312) [inline lib/getWorkflow.js:299]: https://github.com/screwdriver-cd/workflow-parser/pull/43#discussion_r1243970956
`AND` and `OR` sources would be mutually exclusive. We need not skip adding the edge here.
```suggestion
```

**Comment by tkyi** [comment]
Closing in favor of https://github.com/screwdriver-cd/workflow-parser/pull/44

**Pull request closed without merging**